### ISSUE TITLE
Serialization of large trees to protobuf is slow - consider flat representation

### ISSUE THREAD

**Issue by JohnZed** [dmlc/treelite#156]
As per offline discussions, trees of depth 20+ can be very slow to convert to and from protobuf with the current hierarchical format. A float format for the nodes (closer to the native representation) may be significantly faster. A PoC seems worthwhile there.

**Comment by hcho3** [comment]
PoC is available at https://github.com/hcho3/treelite/tree/poc_faster_serializer.

I wrote a [benchmark script](https://gist.github.com/hcho3/f774c512b5f88a2386a4bff0c652593e) to generate a full binary tree with depth = 24. This is written in C to minimize overhead and quickly build 2^24 - 1 nodes.

Results: Flat representation provides 31% speedup over the baseline. Using float32 instead of float64 gives a slight speedup but not much.



  | baseline | flat_protobuf | float_protobuf_float32
-- | -- | -- | --
save | 4.774395 | 3.861472 | 3.791824
load | 12.109890 | 9.040858 | 8.735066
total | 16.884285 | 12.902330 | 12.526891

**Comment by hcho3** [comment]
I implemented binary serialization for the trees, and now performance is significantly better.

Time (sec) | baseline | binary serialization
-- | -- | --
save | 4.774395 | **3.121664**
load | 12.109890 | **2.047686**
total | 16.884285 | **5.169350**

It would be even faster without the two fields `leaf_vector_` and `left_categories_`, both of which are used for LightGBM models.

**Comment by levsnv** [comment]
Re: efficiently supporting multiclass models: since you're willing to break direct protobuf compatibility,
how about this? https://gist.github.com/levsnv/7028c04aa68dee14852e932e6568fdd0
No question that the spec should support all these types itself. This is more to avoid the edge cases and overhead introduced by the possibility of different leaves within the same tree having different data types.

**Comment by hcho3** [comment]
@levsnv I like it. One suggestion is to allow integer output from leaf value even when thresholds are float32. This lets us accommodate multiclass classification with majority voting (See #149).

**Comment by jakirkham** [comment]
IIUC this is solved by PR ( https://github.com/dmlc/treelite/pull/178 ), right?

**Comment by hcho3** [comment]
Yes, it's solved by #178. The new numbers are in https://github.com/hcho3/treelite-dask-serializer.
```
Time to build model: 10.36706407397287 sec
Serialized model to Python buffer frames in 2.8313021175563335e-05 sec
Deserialized model from Python buffer frames in 1.0696996469050646e-05 sec
```

**Comment by jakirkham** [comment]
Awesome! Thanks Philip 😄
